### PR TITLE
Batch Norm rewrite without mshadow, 1D, 2D, 3D, float16, float32, float64 as well as operator gtest framework

### PR DESCRIPTION
**Pull request by cjolivier01**
Note that batch_norm.cu and batch_norm-inl.h are almost entirely new code.
However, github is not rendering the diff by default.  Click "View" so see them.
test_op.h (all new code along with test_util.h and test_perf.h) also is not shown for the same reason.
@piiswrong 

Performance (bs=128, c=3, h=28, w=28):

OLD CPU
BatchNormV1Prop<cpu> 2D:  Timing [Forward] 2828.16 ms, avg: 5.65631 ms X 500 passes
BatchNormV1Prop<cpu> 2D:  Timing [Backward] 20908.4 ms, avg: 41.8169 ms X 500 passes

NEW CPU
BatchNormProp<cpu> 2D:  Timing [Forward] 788.777 ms, avg: 1.57755 ms X 500 passes
BatchNormProp<cpu> 2D:  Timing [Backward] 322.013 ms, avg: 0.644026 ms X 500 passes

OLD GPU
BatchNormV1Prop<gpu> 2D:  Timing [Forward] 5.365 ms, avg: 0.01073 ms X 500 passes
BatchNormV1Prop<gpu> 2D:  Timing [Backward] 15.483 ms, avg: 0.030966 ms X 500 passes

NEW GPU
BatchNormProp<gpu> 2D:  Timing [Forward] 3.514 ms, avg: 0.007028 ms X 500 passes
BatchNormProp<gpu> 2D:  Timing [Backward] 4.787 ms, avg: 0.009574 ms X 500 passes

### CODE REVIEW

File: include/mxnet/tensor_blob.h [hunk 1]
```diff
@@ -184,7 +184,7 @@ class TBlob {
         mshadow::Shape1(shape_.Size()), stream);
   }
   /*! \brief return number of dimension of the tensor inside */
-  inline int ndim(void) const {
+  inline index_t ndim(void) const {
     return shape_.ndim();
```
Comment (piiswrong) [inline include/mxnet/tensor_blob.h]: cast instead of change return type.

Comment (cjolivier01) [inline include/mxnet/tensor_blob.h]: I did this because just about every other ndim() and size() is index_t. This is a weird outlier as int.
Is there a reason it is int?

File: src/operator/operator_common.h [hunk 2]
```diff
@@ -309,6 +309,34 @@ inline void ParamParser(nnvm::NodeAttrs* attrs) {
   attrs->parsed = std::move(param);
 }
 
+/*! \brief Callback class to allow for convenient development and testing */
+template<typename Type>
+class Callbacker {
```
Comment (piiswrong) [inline src/operator/operator_common.h]: ?

Comment (cjolivier01) [inline src/operator/operator_common.h]: For unit testing callback into the unit testing framework. I can move it to just batchnorm if you like.

File: src/operator/batch_norm_v1-inl.h [hunk 3]
```diff
@@ -0,0 +1,366 @@
+/*!
```
Comment (cjolivier01) [inline src/operator/batch_norm_v1-inl.h]: This is basically the old batch_norm-inl.h

File: src/operator/batch_norm_v1.cc [hunk 4]
```diff
@@ -0,0 +1,89 @@
+/*!
```
Comment (cjolivier01) [inline src/operator/batch_norm_v1.cc]: batch_norm_v1.cc is basically the old batch_norm.cc

**Comment by piiswrong** [comment]
please also add python tests.
@reminisce Could you review?

File: include/mxnet/tensor_blob.h (hunk 1 above)
Comment (cjolivier01) [inline include/mxnet/tensor_blob.h]: The size() calls return index_t, btw.

File: src/operator/operator_common.h (hunk 2 above)
Comment (piiswrong) [inline src/operator/operator_common.h]: is a callback necessary? Can you do it in the test code instead?

File: src/operator/batch_norm_v1.cu [hunk 5]
```diff
@@ -0,0 +1,19 @@
+/*!
```
Comment (cjolivier01) [inline src/operator/batch_norm_v1.cu]: batch_norm_v1.cu is basically the old batch_norm.cu

File: tests/jenkins/run_test_ubuntu.sh [hunk 6]
```diff
@@ -19,7 +19,7 @@ fi
 cp make/config.mk .
 echo "USE_CUDA=1" >> config.mk
 echo "USE_CUDA_PATH=/usr/local/cuda" >> config.mk
-echo "USE_CUDNN=1" >> config.mk
+echo "USE_CUDNN=0" >> config.mk
```
Comment (piiswrong) [inline tests/jenkins/run_test_ubuntu.sh]: ?

Comment (cjolivier01) [inline tests/jenkins/run_test_ubuntu.sh]: One the the jenkins scripts should build organic GPU operators rather than CUDNN always

File: src/operator/operator_common.h (hunk 2 above)
Comment (cjolivier01) [inline src/operator/operator_common.h]: The reason that there is a callback call is to inspect intermediate values (print them, for instance) without having to include the test_util.h (for example) in the main build tree.  When not being tested, it has no overhead (production code doesn't make callbacks).

File: tests/jenkins/run_test_ubuntu.sh (hunk 6 above)
Comment (piiswrong) [inline tests/jenkins/run_test_ubuntu.sh]: mkl test is doing that

Comment (cjolivier01) [inline tests/jenkins/run_test_ubuntu.sh]: ok

File: src/operator/batch_norm-inl.h [hunk 7]
```diff
@@ -50,20 +52,294 @@ struct BatchNormParam : public dmlc::Parameter<BatchNormParam> {
...
+        for (size_t i = 0; i < matrixSize; ++i) {
+          onData(channel, in_data++, out_data++);
+        }
+      }
+    }
+  }
+
+  /*! \brief Compute the mean of each input channel */
+  template<typename Shape>
+  static inline void computeMean(const Dtype *in_data,
```
Comment (reminisce) [inline src/operator/batch_norm-inl.h]: Seems you implemented computing mean and variance without using gemm that has been optmized internally. According to your benchmark results, new CPU is slower than the old CPU. Any good reasons?

File: src/operator/batch_norm-inl.h [hunk 8]
```diff
@@ -74,48 +350,172 @@ class BatchNormOp : public Operator {
...
+                  const OpContext &ctx,
+                  const std::vector<TBlob> &out_grad,
+                  const std::vector<TBlob> &in_data,
+                  const std::vector<TBlob> &out_data,
+                  const std::vector<OpReqType> &req,
+                  const std::vector<TBlob> &in_grad,
+                  const std::vector<TBlob> &aux_states);
+#endif  // MXNET_USE_CUDA
+
+  void doBackward(mshadow::Stream<cpu> *stream,
```
Comment (reminisce) [inline src/operator/batch_norm-inl.h]: When MXNET_USE_CUDA=1, this function is compiled. But it shouldn't be, right?

File: src/operator/batch_norm-inl.h [hunk 9]
```diff
@@ -50,20 +52,294 @@ struct BatchNormParam : public dmlc::Parameter<BatchNormParam> {
...
+        for (size_t i = 0; i < matrixSize; ++i) {
+          onData(channel, in_data++);
+        }
+      }
+    }
+  }
+
+  /*! \brief Fast-foreach when you don't care about the position other than channel */
+  template<typename Shape, typename OnData>
+  static inline void forEachFast(Dtype *in_data, const Shape& shape,
```
Comment (yuruofeifei) [inline src/operator/batch_norm-inl.h]: Is there a need for this function?
I think we can just pass the in_data array with correct channel offset.

File: src/operator/batch_norm-inl.h [hunk 10]
```diff
@@ -50,20 +52,294 @@ struct BatchNormParam : public dmlc::Parameter<BatchNormParam> {
...
+                              const Dtype momentum,
+                              const Shape &oshape,
+                              Dtype *save_std) {
+    for (size_t i = 0, n = oshape.Size(); i < n; ++i) {
+      save_std[i] = 0;
+    }
+    const size_t channelCount = ishape[1];
+    CHECK(oshape.Size() == channelCount);
+
+    forEachFast(in_data, ishape,
```
Comment (yuruofeifei) [inline src/operator/batch_norm-inl.h]: How about using std accumulate for mean and variance?
And using std transform for converting data back?

File: src/operator/batch_norm-inl.h (hunk 7 above)
Comment (yuruofeifei) [inline src/operator/batch_norm-inl.h]: How can we apply gemm here?
For mean: we create an array [1/size .... ]T and multiply with the original in_data?
Will gemm consume more space?
Is gemm doing better than openmp?

New to MXNet, thank you for your help.

Comment (reminisce) [inline src/operator/batch_norm-inl.h]: @yuruofeifei Yes. Caffe uses gemm/gemv to implement computing mean and variance. I guess it's faster than putting omp there manually. In addition, CPU and GPU can share the same piece of code if using gemm. Currently, this code is for CPU only. It requires a little more memory to store the vector, but gives faster speed (my guess) and easy code maintenance in return.

File: src/operator/batch_norm-inl.h [hunk 11]
```diff
@@ -50,20 +52,294 @@ struct BatchNormParam : public dmlc::Parameter<BatchNormParam> {
...
 };
 
-template<typename xpu>
-class BatchNormOp : public Operator {
+/*! \brief Batch normalization operator */
+template<typename xpu, typename Dtype, typename AccType>
+class BatchNormOp : public Operator
+                  , public Callbacker<Operator> {
+  typedef ::nnvm::TShape TShape;
+  typedef ::mxnet::TBlob TBlob;
```
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: they get mixed up sometimes with mshadow::TShape/TBlob. They are there because of compile issues.

File: src/operator/batch_norm-inl.h (hunk 7 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: Eric instructed me to not use blas such as caffe does for this, but to base it upon Torch.
In addition, gemm does superfluous multiplies when calculating mean.

File: src/operator/batch_norm-inl.h [hunk 12]
```diff
@@ -50,20 +54,294 @@ struct BatchNormParam : public dmlc::Parameter<BatchNormParam> {
...
+   * \param in_data array of input data, it is const
+   * \param req the request types of saving operation, can only be kWriteTo or kWriteInplace.
+   * \param out_data array of output data, pointer is used to indicate that this is holder
+   *        the space of TBlob in out_data must be pre-allocated with InferShape
+   * \param aux_states Auxiliary states of operator. Normally operator doesn't
+   *        need, epecial case like Batch Norm requires.
+   * \sa OpReqType, OpContext
+   */
+#ifdef MXNET_USE_CUDA
+  void doForward(mshadow::Stream<gpu> *stream,
```
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: ok

File: src/operator/batch_norm-inl.h [hunk 13]
```diff
@@ -50,20 +52,294 @@ struct BatchNormParam : public dmlc::Parameter<BatchNormParam> {
...
+  }
+
+  /*! \brief inverse standard deviation <-> variance
+   * Note that these aren't entirely reversible due to eps
+   **/
+  #define VARIANCE_TO_INVSTD(__var$,    __eps$)   (1.0/sqrt((__var$) + Dtype(__eps$)))
+  #define INVSTD_TO_VARIANCE(__invstd$, __eps$)   ((1.0 / ((__invstd$) * (__invstd$))) - (__eps$))
+
+  /*! \brief Compute the variance of each input channel, as well as update moving mean/variants */
+  template<typename DType, typename  Shape>
```
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: This works with both mshadow::TShape and nnvm::TShape equally without causing extra overhead

File: src/operator/batch_norm-inl.h [hunk 14]
```diff
@@ -50,20 +52,294 @@ struct BatchNormParam : public dmlc::Parameter<BatchNormParam> {
   }
 };
 
-template<typename xpu>
-class BatchNormOp : public Operator {
+/*! \brief Batch normalization operator */
+template<typename xpu, typename Dtype, typename AccType>
+class BatchNormOp : public Operator
+                  , public Callbacker<Operator> {
```
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: Why?

**Comment by cjolivier01** [comment]
Whya re unit tests in python required if I am doing unit tests in C++ (gtest)?  Is it required for all contributors to know python?  What extra is gained by this?

File: include/mxnet/tensor_blob.h (hunk 1 above)
Comment (sxjscience) [inline include/mxnet/tensor_blob.h]: Other codes may rely on the return type of ndim() and we should check more carefully before making the change.

File: src/operator/batch_norm-inl.h (hunk 8 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: It is required.  One can always select not to use a GPU even when the code is compiled with GPU enabled

File: src/operator/batch_norm-inl.h (hunk 9 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: It is my understanding that data is arranged like this:

[batch item 1][channel 1][spatial data (d,r,w)][channel 2][spatial data (d,r,w)][batch item 2]
[batch item 2][channel 1][spatial data (d,r,w)][channel 2][spatial data (d,r,w)][batch item 2]

In which case you still need to loop across the batch item, correct?

File: src/operator/batch_norm-inl.h (hunk 10 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: 1) Same question as above
2) Once the omp loop is inserted across channels, then there's not much left to insert into binary operation. In addition, this adds a lot of overhead in debug mode with function calls for what's a pretty simple calculation.
wdyt?

File: src/operator/batch_norm.cu [hunk 15]
```diff
@@ -1,15 +1,541 @@
...
+    }
+  }
+  return DeviceTensor<Dtype, Dim>(data, &size[0]);
+}
+
+#define DeviceTensor1 DeviceTensor<Dtype, 1>
+#define DeviceTensor3 DeviceTensor<Dtype, 3>
+
+template <typename Dtype, typename accreal>
+static void BatchNormalization_updateOutput(mshadow::Stream<gpu> *s,
```
Comment (cjolivier01) [inline src/operator/batch_norm.cu]: ok

File: src/operator/batch_norm-inl.h (hunk 10 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: I changed the invstd->var to transform

File: include/mxnet/tensor_blob.h (hunk 1 above)
Comment (cjolivier01) [inline include/mxnet/tensor_blob.h]: ok

File: src/operator/batch_norm-inl.h (hunk 14 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: It is convenient because I can use dynamic_cast in the unit tests to check for existence of the interface.  Otherwise, for other operators, a change would have to be made to add a function returning a pointer to the callback interface, etc.

File: src/operator/operator_common.h (hunk 2 above)
Comment (piiswrong) [inline src/operator/operator_common.h]: ok let's move this to batchnorm, or better, remove this and only test public methods so that no callback is needed.

File: .gitignore [hunk 16]
```diff
@@ -121,10 +119,10 @@ scala-package/*/*/target/
 .pydevproject
 CMakeFiles
 cmake_install.cmake
-dmlc-core
-ps-lite
-nnvm
-!src/nnvm
+#dmlc-core
```
Comment (piiswrong) [inline .gitignore]: ？

File: src/common/cuda_utils.h [hunk 17]
```diff
@@ -11,6 +11,23 @@
 
 #if MXNET_USE_CUDA
 
+/*! \brief Macros/inlines to assist CLion to parse Cuda files (*.cu, *.cuh) */
+#ifdef __JETBRAINS_IDE__
```
Comment (piiswrong) [inline src/common/cuda_utils.h]: ？

File: src/operator/batch_norm-inl.h [hunk 18]
```diff
@@ -74,48 +353,172 @@ class BatchNormOp : public Operator {
...
+                  const std::vector<TBlob> &out_grad,
+                  const std::vector<TBlob> &in_data,
+                  const std::vector<TBlob> &out_data,
+                  const std::vector<OpReqType> &req,
+                  const std::vector<TBlob> &in_grad,
+                  const std::vector<TBlob> &aux_states);
+#endif  // MXNET_USE_CUDA
+
+  void DoBackward(mshadow::Stream<cpu> *stream,
+                  const OpContext &ctx,
```
Comment (piiswrong) [inline src/operator/batch_norm-inl.h]: declare with stream<xpu>  here and move implementation to .cc
also move the static inline functions to .cc and put them in batchnorm namespace as global functions.

File: src/operator/operator_common.h (hunk 2 above)
Comment (cjolivier01) [inline src/operator/operator_common.h]: ok

File: .gitignore (hunk 16 above)
Comment (cjolivier01) [inline .gitignore]: Why would these be ignored? They are submodules. CLion gives warnings about this and treats the subdirectories strangely.

File: src/common/cuda_utils.h (hunk 17 above)
Comment (cjolivier01) [inline src/common/cuda_utils.h]: This allows CLion IDE to parse cuda code.  It has no net effect on anything else.

File: src/operator/batch_norm-inl.h (hunk 18 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: ok

Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: ok.  Curious, What's the advantage?

Comment (piiswrong) [inline src/operator/batch_norm-inl.h]: looks leaner. cuda implementation is in .cu and cpu implementation is in .cc

**Comment by cjolivier01** [comment]
This change has a build dependency on this PR:
https://github.com/dmlc/mshadow/pull/236

File: src/operator/batch_norm-inl.h [hunk 19]
```diff
@@ -14,25 +14,29 @@
...
 }  // namespace batchnorm
 
+/*! \brief Parameters for BatchNoram operator */
 struct BatchNormParam : public dmlc::Parameter<BatchNormParam> {
   float eps;
   float momentum;
   bool fix_gamma;
   bool use_global_stats;
   bool output_mean_var;
+  int channel_position;
```
Comment (piiswrong) [inline src/operator/batch_norm-inl.h]: axis

Comment (piiswrong) [inline src/operator/batch_norm-inl.h]: https://keras.io/layers/normalization/#batchnormalization

File: tests/python/gpu/test_operator_gpu.py [hunk 20]
```diff
@@ -236,13 +236,99 @@ def test_fft():
             check_fft(shape)
 
 def test_batchnorm_with_type():
-    sym = mx.sym.BatchNorm(name='norm', fix_gamma=False)
-    ctx_list = [{'ctx': mx.gpu(0), 'norm_data': (10, 2, 10, 10), 'type_dict': {'norm_data': np.float32}},
-                {'ctx': mx.cpu(0), 'norm_data': (10, 2, 10, 10), 'type_dict': {'norm_data': np.float32}}]
-    check_consistency(sym, ctx_list)
+  ctx_list = [{'ctx': mx.gpu(0), 'norm_data': (10, 2, 10, 10), 'type_dict': {'norm_data': np.float32}},
```
Comment (piiswrong) [inline tests/python/gpu/test_operator_gpu.py]: add more types since we support it now

File: tests/python/gpu/test_operator_gpu.py [hunk 21]
```diff
@@ -236,13 +236,99 @@ def test_fft():
...
+                                   fix_gamma=fix_gamma, use_global_stats=use_global_stats)
+
+  def test_3d_batchnorm(fix_gamma, use_global_stats):
+    data = (2, 3, 3, 5, 5)
+    test_batchnorm_versions_helper(batchnorm_op_list=['batchnorm_cpu',
+                                                      'batchnorm_gpu'],
+                                   data=data,
+                                   fix_gamma=fix_gamma, use_global_stats=use_global_stats)
+
+  test_1d_batchnorm(True,  False)
```
Comment (piiswrong) [inline tests/python/gpu/test_operator_gpu.py]: test channel positions

Comment (piiswrong) [inline tests/python/gpu/test_operator_gpu.py]: also test cudnn vs cudnn_off

Comment (cjolivier01) [inline tests/python/gpu/test_operator_gpu.py]: channel_position temporarily reverted (going to put in another PR)
cudnn_off is being tested

File: src/operator/batch_norm.cc [hunk 22]
```diff
@@ -2,41 +2,462 @@
...
+  const AccReal  *b = bias.dptr<AccReal>();
+
+  // optionally, keep weights fixed at 1
+  if (param_.fix_gamma) {
+    for (size_t i =0, n = weights.Size(); i < n; ++i) {
+      w[i] = AccReal(1);
+    }
+  }
+
+  if (req[batchnorm::kData] == kWriteTo || req[batchnorm::kData] == kWriteInplace) {
```
Comment (yuruofeifei) [inline src/operator/batch_norm.cc]: nitpick, can use your IsWriting helper function

Comment (cjolivier01) [inline src/operator/batch_norm.cc]: ok

File: .gitignore (hunk 16 above)
Comment (cjolivier01) [inline .gitignore]: why would submodules be ignored?

Comment (cjolivier01) [inline .gitignore]: I put those there accidentally in the caffe data iter commit

**Comment by cjolivier01** [comment]
still pending: req on gamma and beta for forward pass

**Comment by cjolivier01** [comment]
Ok, something I was stuck on that I learned: There is a limit to the number of parameters you can pass the cuda execution.  At some point, you'll get too many resources requested error.

**Comment by piiswrong** [comment]
You can use launch bound

**Comment by cjolivier01** [comment]
Failure does not appear to be related to my code. Merging in latest...

**Comment by cjolivier01** [comment]
Strange new compile problem with R, requires quiet = FALSE to be part of installing roxygen command.  Also occurs on master branch.  . Nothing to do with my code, but pushed fix here.

File: .gitignore (hunk 16 above)
Comment (piiswrong) [inline .gitignore]: are you removing this?

File: src/operator/batch_norm-inl.h [hunk 23]
```diff
@@ -74,154 +103,96 @@ class BatchNormOp : public Operator {
...
-             broadcast<1>(
-                 1.0f / F<mshadow_op::square_root>(moving_var + param_.eps), data.shape_));
-    }
+    mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+    DoBackward(s, ctx, out_grad, in_data,
+               out_data, req, in_grad, aux_states);
   }
 
  private:
+  void DoForward(mshadow::Stream<cpu> *stream,
```
Comment (piiswrong) [inline src/operator/batch_norm-inl.h]: this should be <xpu> and the MXNET_USE_CUDA section can be removed.

File: src/operator/batch_norm-inl.h [hunk 24]
```diff
@@ -348,11 +325,21 @@ class BatchNormProp : public OperatorProperty {
   Operator* CreateOperatorEx(Context ctx, std::vector<TShape> *in_shape,
       std::vector<int> *in_type) const override;
 
+  inline const BatchNormParam& getParam() const {
```
Comment (piiswrong) [inline src/operator/batch_norm-inl.h]: why?

File: src/operator/batch_norm.cc [hunk 25]
```diff
@@ -2,41 +2,492 @@
...
+                      *out_data = static_cast<DType>(
+                        ((*in_data - mean[channel]) * var[channel]) + b[channel]);
+                    });
+      }
+    }
+
+  // Convert back to "real" variance in order to be consistent
+  // with the original operator
+  for (size_t i = 0, n = inputData.shape_[1]; i < n; ++i) {
+    var[i] = INVSTD_TO_VARIANCE(var[i], param_.eps);
```
Comment (piiswrong) [inline src/operator/batch_norm.cc]: These conversions can lose precision and cause underflow problems. Why not store it as variance in the first place?

File: src/operator/batch_norm.cc [hunk 26]
```diff
@@ -2,41 +2,492 @@
...
-    DO_BIND_DISPATCH(CreateOp, param_, (*in_type)[0]);
+                                          std::vector<int> *in_type) const {
+  std::vector<TShape> out_shape, aux_shape;
+  std::vector<int> out_type, aux_type;
+  CHECK(InferType(in_type, &out_type, &aux_type));
+  CHECK(InferShape(in_shape, &out_shape, &aux_shape));
+  BatchNormParam param = param_;
+  CHECK_GE(in_shape->size(), 1U);
+  if ((*in_shape)[0].ndim() != 4) {
+    param.mkl_off = true;
```
Comment (piiswrong) [inline src/operator/batch_norm.cc]: This should be done in CreateOp. You don't need mkl_off param.

File: src/operator/batch_norm.cc [hunk 27]
```diff
@@ -77,21 +528,23 @@ Both ``gamma`` and ``beta`` are learnable parameters. But if ``fix_gamma`` is tr
 then set ``gamma`` to 1 and its gradient to 0.
 
 )code" ADD_FILELINE)
-.add_argument("data", "NDArray-or-Symbol", "Input data to batch normalization")
-.add_argument("gamma", "NDArray-or-Symbol", "gamma array")
-.add_argument("beta", "NDArray-or-Symbol", "beta array")
-.add_arguments(BatchNormParam::__FIELDS__());
+  .add_argument("data", "NDArray-or-Symbol", "Input data to batch normalization")
```
Comment (piiswrong) [inline src/operator/batch_norm.cc]: Don't change this kind of formatting.
It was consistent with the format elsewhere. Your new format is not.
Also it introduces unnecessary merge conflicts for other people

File: src/operator/batch_norm.cu [hunk 28]
```diff
@@ -1,32 +1,670 @@
...
+  }
+  return maxThreadCount;
+}
+
+static const unsigned WARP_SIZE = 32;
+
+// The maximum number of threads in a block
+static const unsigned MAX_BLOCK_SIZE = 512U;
+
+static const unsigned DYN_MAX_BLOCK_SIZE = std::min(GetMaxThreadCount(false), MAX_BLOCK_SIZE);
```
Comment (piiswrong) [inline src/operator/batch_norm.cu]: Don't do this. It forces the intialization of cuda driver at start up, even when users are not actually using GPU.

Use the global default in mshadow constants

Comment (cjolivier01) [inline src/operator/batch_norm.cu]: ok

File: src/operator/batch_norm.cc (hunk 25 above)
Comment (cjolivier01) [inline src/operator/batch_norm.cc]: 1) To be consistent with V1
2) It's used in this format during the backprop for each data point calculation, so this avoids recalculating variance from invstd for each data point pass (which is one of the things that makes the mshadow version slow).

File: src/operator/batch_norm.cc (hunk 26 above)
Comment (cjolivier01) [inline src/operator/batch_norm.cc]: I don't have access to the shape in CreateOp() in order to determine this. I also figured this was useful for the MKLML problem that we wer ehaving earlier where a certain MKL operator had numerical issues and we had to disable all of MKL rather than having an option to just turn off the one problematic operator (without a recompile).

File: src/operator/batch_norm.cc (hunk 27 above)
Comment (cjolivier01) [inline src/operator/batch_norm.cc]: I think this was done by the IDE.  I will revert.

File: src/operator/batch_norm.cc (hunk 25 above)
Comment (cjolivier01) [inline src/operator/batch_norm.cc]: Let me check that it isn't cycling the calculation

File: src/operator/batch_norm.cu (hunk 28 above)
Comment (cjolivier01) [inline src/operator/batch_norm.cu]: whats that var?

File: src/operator/batch_norm.cc (hunk 26 above)
Comment (cjolivier01) [inline src/operator/batch_norm.cc]: Am i permitted to change this createop signature? it seems pretty standardize right now.

File: src/operator/batch_norm-inl.h (hunk 24 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: For the unit testing

File: src/operator/batch_norm-inl.h (hunk 23 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: sorry, I don't understand this statement. should be xpu?

File: .gitignore (hunk 16 above)
Comment (cjolivier01) [inline .gitignore]: yes, removed

File: src/operator/batch_norm-inl.h (hunk 23 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: As far as I can find, it is not possible to partially specialize the template parameters out-of-line this way (i.e. only specify xpu and not DType and AccReal in the specialization).

Note: 

http://stackoverflow.com/questions/3768862/c-single-template-specialisation-with-multiple-template-parameters

Which would be:  

http://stackoverflow.com/questions/3768862/c-single-template-specialisation-with-multiple-template-parameters

For example:

template <typename F, typename I>
class B
{
 public:
  int x_;
  void someFunc()
  {
    std::cout << "normal" << std::endl;
  };

  void someFuncNotSpecial()
  {
    std::cout << "normal" << std::endl;
  };
};

template <typename I>
class B<float, I>
{
 public:
  void someFunc();
};

template <typename I>
void B<float, I>::someFunc()
{
  x_ = 0;
}

This will compile with x_ not found.

File: src/operator/batch_norm.cc (hunk 25 above)
Comment (cjolivier01) [inline src/operator/batch_norm.cc]: For performance, in forward pass, invstd is part of the per-item calculation.  Also, this is consistent with V1 for unit testing.
mshadow (V1) version converts liberally between the two.
To avoid possible cycling, I will adjust the forward pass a little to not convert the same value both ways

File: src/operator/batch_norm.cc (hunk 26 above)
Comment (piiswrong) [inline src/operator/batch_norm.cc]: you can feed shape and type to createop. See other ops for example.

keep adding xxx_off isn't a sustainable solution. We should have something generic in the future. Until then please don't add new stuff.

**Comment by piiswrong** [comment]
one last change in tests: initialize gamma and beta to random (instead of ones and zeros) to make sure that fix_gamma is working the same way as before

**Comment by cjolivier01** [comment]
Already checking varying gamma/beta per our offline discussion

File: src/operator/batch_norm.cc [hunk 29]
```diff
@@ -2,35 +2,486 @@
...
+  CHECK_GE(in_shape->size(), 1U);
+  const int dim = (*in_shape)[0].ndim();
+  if (dim != 4) {
+    param.mkl_off = true;
+    if (dim > 4) {
+      param.cudnn_off = true;
+    }
+  }
+
+  DO_BIND_DISPATCH(CreateOp, param, (*in_type)[0]);
```
Comment (piiswrong) [inline src/operator/batch_norm.cc]: pass in_shape to createop here and remove mkl_off

Comment (cjolivier01) [inline src/operator/batch_norm.cc]: ok
